File: src/lava/magma/runtime/message_infrastructure/test/test_shm_delivery.py
# Copyright (C) 2021 Intel Corporation
# SPDX-License-Identifier: BSD-3-Clause
# See: https://spdx.org/licenses/

import numpy as np
import unittest
from functools import partial
import time
from datetime import datetime
from multiprocessing import shared_memory
from multiprocessing import Semaphore
from multiprocessing import Process

from message_infrastructure.multiprocessing import MultiProcessing

from message_infrastructure import (
    ChannelBackend,
    Channel,
)


class process():

    def __init__(self) -> None:
        pass

    def get_status(self):
        return 0


class PyChannel():

    def __init__(self, dtype, size, nbytes, name, *_) -> None:
        self.shm_ = Shm(dtype, size, nbytes, name)
        self.src_port = Port(self.shm_)
        self.dst_port = Port(self.shm_)
        self.dtype_ = dtype
        self.shm_.start()

    def start(self):
        pass

    def join(self):
        pass


class Port():

    def __init__(self, shm) -> None:
        self.shm_ = shm

    def send(self, data):
        return self.shm_.push(data)

    def recv(self):
        return self.shm_.pop()

    def start(self):
        pass

    def join(self):
        pass


class Shm():
    def __init__(self, dtype, size, nbytes, name) -> None:
        self.shm_ = shared_memory.SharedMemory(name=name,
                                               create=True, size=nbytes * size)
        self.nbytes_ = nbytes
        self.size_ = size
        self.sem_ack_ = Semaphore(size)
        self.sem_req_ = Semaphore(0)
        self.sem_ = Semaphore(0)
        self.read_ = 0
        self.write_ = 0
        self.type_ = dtype

    def push(self, data):
        self.sem_ack_.acquire()
        self.sem_.acquire()
        self.shm_.buf[self.write_ * self.nbytes_:
                      ((self.write_ + 1) * self.nbytes_)] = bytearray(data)
        self.write_ = (self.write_ + 1) % self.size_
        self.type_ = data.dtype
        self.sem_.release()
        self.sem_req_.release()

    def pop(self):
        self.sem_req_.acquire()
        self.sem_.acquire()
        result = bytearray(self.shm_.buf[self.read_ * self.nbytes_:
                           ((self.read_ + 1) * self.nbytes_)])
        self.read_ = (self.read_ + 1) % self.size_
        self.sem_.release()
        self.sem_ack_.release()
        return np.frombuffer(result, self.type_)

    def start(self):
        self.sem_.release()

    def __del__(self):
        self.shm_.close()
        self.shm_.unlink()


class Builder():
    def build(self, i):
        pass


def bound_target_a1(loop, mp_to_a1, a1_to_a2,
                    a2_to_a1, a1_to_mp, this, builder):
    from_mp = mp_to_a1.dst_port
    from_mp.start()
    to_a2 = a1_to_a2.src_port
    to_a2.start()
    from_a2 = a2_to_a1.dst_port
    from_a2.start()
    to_mp = a1_to_mp.src_port
    to_mp.start()
    while loop > 0 and this.get_status() == 0:
        loop = loop - 1
        data = from_mp.recv()
        data[0] = data[0] + 1
        to_a2.send(data)
        data = from_a2.recv()
        data[0] = data[0] + 1
        to_mp.send(data)

    while this.get_status() == 0:
        time.sleep(0.0001)
    from_mp.join()
    to_a2.join()
    from_a2.join()
    to_mp.join()


def bound_target_a2(loop, a1_to_a2, a2_to_a1, this, builder):
    from_a1 = a1_to_a2.dst_port
    from_a1.start()
    to_a1 = a2_to_a1.src_port
    to_a1.start()
    while loop > 0 and this.get_status() == 0:
        loop = loop - 1
        data = from_a1.recv()
        data[0] = data[0] + 1
        to_a1.send(data)

    while this.get_status() == 0:
        time.sleep(0.0001)
    from_a1.join()
    to_a1.join()


def prepare_data():
    return np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 0] * 1000)


class TestShmDelivery(unittest.TestCase):

    def __init__(self, methodName: str = ...) -> None:
        super().__init__(methodName)
        self.loop_ = 100000


    def test_cpp_shm_loop_with_cpp_multiprocess(self):
        loop = self.loop_
        mp = MultiProcessing()
        mp.start()
        predata = prepare_data()
        queue_size = 2
        nbytes = np.prod(predata.shape) * predata.dtype.itemsize
        mp_to_a1 = Channel(
            ChannelBackend.SHMEMCHANNEL,
            queue_size,
            nbytes,
            "mp_to_a1",
            "mp_to_a1")
        a1_to_a2 = Channel(
            ChannelBackend.SHMEMCHANNEL,
            queue_size,
            nbytes,
            "a1_to_a2",
            "a1_to_a2")
        a2_to_a1 = Channel(
            ChannelBackend.SHMEMCHANNEL,
            queue_size,
            nbytes,
            "a2_to_a1",
            "a2_to_a1")
        a1_to_mp = Channel(
            ChannelBackend.SHMEMCHANNEL,
            queue_size,
            nbytes,
            "a1_to_mp",
            "a1_to_mp")

        target_a1 = partial(bound_target_a1, loop, mp_to_a1,
                            a1_to_a2, a2_to_a1, a1_to_mp)
        target_a2 = partial(bound_target_a2, loop, a1_to_a2, a2_to_a1)

        builder = Builder()

        mp.build_actor(target_a1, builder)  # actor1
        mp.build_actor(target_a2, builder)  # actor2

        to_a1 = mp_to_a1.src_port
        from_a1 = a1_to_mp.dst_port

        to_a1.start()
        from_a1.start()

        expect_result = np.copy(predata)
        expect_result[0] = (1 + 3 * loop)
        loop_start = datetime.now()
        while loop > 0:
            loop = loop - 1
            print(loop)
            to_a1.send(predata)
            predata = from_a1.recv()
        loop_end = datetime.now()

        if not np.array_equal(expect_result, predata):
            print("expect: ", expect_result)
            print("result: ", predata)
            raise AssertionError()

        to_a1.join()
        from_a1.join()
        mp.stop(True)
        print("cpp_shm_loop_with_cpp_multiprocess timedelta =",
              loop_end - loop_start)


    def test_cpp_skt_loop_with_cpp_multiprocess(self):
        loop = self.loop_
        mp = MultiProcessing()
        mp.start()
        predata = prepare_data()
        queue_size = 2
        nbytes = np.prod(predata.shape) * predata.dtype.itemsize
        mp_to_a1 = Channel(
            ChannelBackend.SOCKETCHANNEL,
            queue_size,
            nbytes,
            "mp_to_a1",
            "mp_to_a1")
        a1_to_a2 = Channel(
            ChannelBackend.SOCKETCHANNEL,
            queue_size,
            nbytes,
            "a1_to_a2",
            "a1_to_a2")
        a2_to_a1 = Channel(
            ChannelBackend.SOCKETCHANNEL,
            queue_size,
            nbytes,
            "a2_to_a1",
            "a2_to_a1")
        a1_to_mp = Channel(
            ChannelBackend.SOCKETCHANNEL,
            queue_size,
            nbytes,
            "a1_to_mp",
            "a1_to_mp")

        target_a1 = partial(bound_target_a1, loop, mp_to_a1,
                            a1_to_a2, a2_to_a1, a1_to_mp)
        target_a2 = partial(bound_target_a2, loop, a1_to_a2, a2_to_a1)

        builder = Builder()

        mp.build_actor(target_a1, builder)  # actor1
        mp.build_actor(target_a2, builder)  # actor2

        to_a1 = mp_to_a1.src_port
        from_a1 = a1_to_mp.dst_port

        to_a1.start()
        from_a1.start()

        expect_result = np.copy(predata)
        expect_result[0] = (1 + 3 * loop)
        loop_start = datetime.now()
        while loop > 0:
            loop = loop - 1
            print(loop)
            to_a1.send(predata)
            predata = from_a1.recv()
        loop_end = datetime.now()

        if not np.array_equal(expect_result, predata):
            print("expect: ", expect_result)
            print("result: ", predata)
            raise AssertionError()

        to_a1.join()
        from_a1.join()
        mp.stop(True)
        print("cpp_skt_loop_with_cpp_multiprocess timedelta =",
              loop_end - loop_start)


    def test_py_shm_loop_with_cpp_multiprocess(self):
        loop = self.loop_

        mp = MultiProcessing()
        mp.start()

        predata = prepare_data()
        queue_size = 2
        nbytes = np.prod(predata.shape) * predata.dtype.itemsize
        mp_to_a1 = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "mp_to_a1",
            "mp_to_a1")
        a1_to_a2 = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "a1_to_a2",
            "a1_to_a2")
        a2_to_a1 = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "a2_to_a1",
            "a2_to_a1")
        a1_to_mp = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "a1_to_mp",
            "a1_to_mp")

        builder = Builder()

        target_a1 = partial(bound_target_a1, loop, mp_to_a1,
                            a1_to_a2, a2_to_a1, a1_to_mp)
        target_a2 = partial(bound_target_a2, loop, a1_to_a2, a2_to_a1)

        mp.build_actor(target_a1, builder)  # actor1
        mp.build_actor(target_a2, builder)  # actor2

        to_a1 = mp_to_a1.src_port
        from_a1 = a1_to_mp.dst_port

        to_a1.start()
        from_a1.start()

        expect_result = np.copy(predata)
        expect_result[0] = (1 + 3 * loop)

        loop_start = datetime.now()
        while loop > 0:
            loop = loop - 1
            to_a1.send(predata)
            predata = from_a1.recv()
        loop_end = datetime.now()

        if not np.array_equal(expect_result, predata):
            print("expect: ", expect_result)
            print("result: ", predata)
            raise AssertionError()

        to_a1.join()
        from_a1.join()
        mp.stop(True)
        print("py_shm_loop_with_cpp_multiprocess timedelta =",
              loop_end - loop_start)


    def test_py_shm_loop_with_py_multiprocess(self):
        loop = self.loop_

        predata = prepare_data()
        queue_size = 2
        nbytes = np.prod(predata.shape) * predata.dtype.itemsize
        mp_to_a1 = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "mp_to_a1",
            "mp_to_a1")
        a1_to_a2 = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "a1_to_a2",
            "a1_to_a2")
        a2_to_a1 = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "a2_to_a1",
            "a2_to_a1")
        a1_to_mp = PyChannel(
            predata.dtype,
            queue_size,
            nbytes,
            "a1_to_mp",
            "a1_to_mp")

        builder = Builder()

        target_a1 = partial(bound_target_a1, loop, mp_to_a1,
                            a1_to_a2, a2_to_a1, a1_to_mp, process(), builder)
        target_a2 = partial(bound_target_a2, loop, a1_to_a2, a2_to_a1,
                            process(), builder)

        a1 = Process(target=target_a1)
        a2 = Process(target=target_a2)
        a1.start()
        a2.start()

        to_a1 = mp_to_a1.src_port
        from_a1 = a1_to_mp.dst_port

        to_a1.start()
        from_a1.start()

        expect_result = np.copy(predata)
        expect_result[0] = (1 + 3 * loop)

        loop_start = datetime.now()
        while loop > 0:
            loop = loop - 1
            to_a1.send(predata)
            predata = from_a1.recv()
        loop_end = datetime.now()

        if not np.array_equal(expect_result, predata):
            print("expect: ", expect_result)
            print("result: ", predata)
            raise AssertionError()

        to_a1.join()
        from_a1.join()
        a1.terminate()
        a2.terminate()
        a1.join()
        a2.join()
        print("py_shm_loop_with_py_multiprocess timedelta =",
              loop_end - loop_start)


if __name__ == '__main__':
    unittest.main()

File: src/lava/magma/runtime/message_infrastructure/test/cpptest/test_shm_delivery.cc
// Copyright (C) 2022 Intel Corporation
// SPDX-License-Identifier: BSD-3-Clause
// See: https://spdx.org/licenses/

#include <message_infrastructure/csrc/core/channel_factory.h>
#include <message_infrastructure/csrc/core/multiprocessing.h>
#include <message_infrastructure/csrc/channel/shmem/shmem_channel.h>
#include <message_infrastructure/csrc/core/message_infrastructure_logging.h>
#include <gtest/gtest.h>
#include <cstring>

namespace message_infrastructure {

void stop_fn() {
  // exit(0);
}

void target_fn_a1_bound(
  int loop,
  AbstractChannelPtr mp_to_a1,
  AbstractChannelPtr a1_to_mp,
  AbstractChannelPtr a1_to_a2,
  AbstractChannelPtr a2_to_a1,
  AbstractActor* actor_ptr) {
    actor_ptr->SetStopFn(stop_fn);
    auto from_mp = mp_to_a1->GetRecvPort();
    from_mp->Start();
    auto to_mp   = a1_to_mp->GetSendPort();
    to_mp->Start();
    auto to_a2   = a1_to_a2->GetSendPort();
    to_a2->Start();
    auto from_a2 = a2_to_a1->GetRecvPort();
    from_a2->Start();
    // LAVA_DUMP(1, "shm actor1, loop: %d\n", loop);
    while ((loop--)&&!actor_ptr->GetStatus()) {
      // LAVA_DUMP(1, "shm actor1 waitting\n");
      MetaDataPtr data = from_mp->Recv();
      // LAVA_DUMP(1, "shm actor1 recviced\n");
      (*reinterpret_cast<int64_t*>(data->mdata))++;
      to_a2->Send(data);
      free(reinterpret_cast<char*>(data->mdata));
      data = from_a2->Recv();
      (*reinterpret_cast<int64_t*>(data->mdata))++;
      to_mp->Send(data);
      free(reinterpret_cast<char*>(data->mdata));
    }
    from_mp->Join();
    from_a2->Join();
    while (!actor_ptr->GetStatus()) {
      helper::Sleep();
    }
  }

void target_fn_a2_bound(
  int loop,
  AbstractChannelPtr a1_to_a2,
  AbstractChannelPtr a2_to_a1,
  AbstractActor* actor_ptr) {
    actor_ptr->SetStopFn(stop_fn);
    auto from_a1 = a1_to_a2->GetRecvPort();
    from_a1->Start();
    auto to_a1   = a2_to_a1->GetSendPort();
    to_a1->Start();
    // LAVA_DUMP(1, "shm actor2, loop: %d\n", loop);
    while ((loop--)&&!actor_ptr->GetStatus()) {
      // LAVA_DUMP(1, "shm actor2 waitting\n");
      MetaDataPtr data = from_a1->Recv();
      // LAVA_DUMP(1, "shm actor2 recviced\n");
      (*reinterpret_cast<int64_t*>(data->mdata))++;
      to_a1->Send(data);
      free(reinterpret_cast<char*>(data->mdata));
    }
    from_a1->Join();
    while (!actor_ptr->GetStatus()) {
      helper::Sleep();
    }
  }

TEST(TestShmDelivery, ShmLoop) {
  MultiProcessing mp;
  int loop = 10000;
  const int queue_size = 1;
  AbstractChannelPtr mp_to_a1 = GetChannelFactory().GetChannel(
    SHMEMCHANNEL, queue_size, sizeof(int64_t)*10000, "mp_to_a1", "mp_to_a1");
  AbstractChannelPtr a1_to_mp = GetChannelFactory().GetChannel(
    SHMEMCHANNEL, queue_size, sizeof(int64_t)*10000, "a1_to_mp", "a1_to_mp");
  AbstractChannelPtr a1_to_a2 = GetChannelFactory().GetChannel(
    SHMEMCHANNEL, queue_size, sizeof(int64_t)*10000, "a1_to_a2", "a1_to_a2");
  AbstractChannelPtr a2_to_a1 = GetChannelFactory().GetChannel(
    SHMEMCHANNEL, queue_size, sizeof(int64_t)*10000, "a2_to_a1", "a2_to_a1");
  auto target_fn_a1 = std::bind(&target_fn_a1_bound, loop,
                                mp_to_a1, a1_to_mp, a1_to_a2,
                                a2_to_a1, std::placeholders::_1);
  auto target_fn_a2 = std::bind(&target_fn_a2_bound, loop, a1_to_a2,
                                a2_to_a1, std::placeholders::_1);
  int actor1 = mp.BuildActor(target_fn_a1);
  int actor2 = mp.BuildActor(target_fn_a2);
  auto to_a1   = mp_to_a1->GetSendPort();
  to_a1->Start();
  auto from_a1 = a1_to_mp->GetRecvPort();
  from_a1->Start();
  int64_t array_[10000] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 0};
  std::fill(array_ + 10, array_ + 10000, 1);

  MetaDataPtr metadata = std::make_shared<MetaData>();
  metadata->nd = 1;
  metadata->type = 7;
  metadata->elsize = 8;
  metadata->total_size = 10000;
  metadata->dims[0] = 10000;
  metadata->strides[0] = 1;
  metadata->mdata =
    reinterpret_cast<char*>
    (malloc(sizeof(int64_t)*10000));
  std::memcpy(metadata->mdata,
              reinterpret_cast<char*>(array_),
              metadata->elsize * metadata->total_size);
  MetaDataPtr mptr;
  LAVA_DUMP(1, "main process loop: %d\n", loop);
  int expect_result = 1 + loop * 3;
  const clock_t start_time = std::clock();
  while (loop--) {
    to_a1->Send(metadata);
    free(reinterpret_cast<char*>(metadata->mdata));
    // LAVA_DUMP(1, "shm wait for response, remain loop: %d\n", loop);
    mptr = from_a1->Recv();
    // to_a1->Join();
    // LAVA_DUMP(1, "metadata:\n");
    // LAVA_DUMP(1, "nd: %ld\n", mptr->nd);
    // LAVA_DUMP(1, "type: %ld\n", mptr->type);
    // LAVA_DUMP(1, "elsize: %ld\n", mptr->elsize);
    // LAVA_DUMP(1, "total_size: %ld\n", mptr->total_size);
    // LAVA_DUMP(1, "dims: {%ld, %ld, %ld, %ld, %ld}\n",
              // mptr->dims[0], mptr->dims[1], mptr->dims[2],
              // mptr->dims[3], mptr->dims[4]);
    // LAVA_DUMP(1, "strides: {%ld, %ld, %ld, %ld, %ld}\n",
    // mptr->strides[0], mptr->strides[1], mptr->strides[2], mptr->strides[3],
    // mptr->strides[4]);
    // LAVA_DUMP(1, "mdata: %p, *mdata: %ld\n", mptr->mdata,
    //           *reinterpret_cast<int64_t*>(mptr->mdata));
    int64_t *ptr = reinterpret_cast<int64_t*>(mptr->mdata);
    // for (int i = 0; i < 20; i++) {
    //   LAVA_DUMP(1, "shm mdata: %p, shm *mdata: %ld\n", ptr, *ptr);
    //   ptr++;
    // }
    metadata = mptr;
  }
  const clock_t end_time = std::clock();
  int64_t result = *reinterpret_cast<int64_t*>(metadata->mdata);
  printf("shm result =%ld", result);
  free(reinterpret_cast<char*>(mptr->mdata));
  from_a1->Join();
  mp.Stop(true);
  if (result != expect_result) {
    LAVA_DUMP(1, "expect_result: %d\n", expect_result);
    LAVA_DUMP(1, "result: %ld\n", result);
    LAVA_LOG_ERR("result != expect_result");
    throw;
  }
  printf("shm cpp loop timedelta: %f s\n",
        ((end_time - start_time)/static_cast<double>(CLOCKS_PER_SEC)));
  LAVA_DUMP(1, "shm cpp loop timedelta: %f",
           ((end_time - start_time)/static_cast<double>(CLOCKS_PER_SEC)));
  LAVA_DUMP(1, "exit\n");
}

}  // namespace message_infrastructure
